python/cucim/src/cucim/core/operations/color/jitter.py

```diff
@@ -0,0 +1,300 @@
+    saturation : float or 2-tuple of float, optional
+        Non-negative factor to jitter the saturation by. When `saturation` is a
+        scalar, scaling will be by a random value between
+        ``[max(0, 1 - saturation), (1 + saturation)]``. `saturation` can
+        also be a 2-tuple specifying the range for the random scaling factor.
+        A value of 0 or (1, 1) will result in no change.
+    hue : float or 2-tuple of float, optional
+        Factor between [-0.5, 0.5] to jitter hue by. When `hue` is a
+        scalar, scaling will be by a random value between in the range
+        ``[-hue, hue]``. `hue` can also be a 2-tuple specifying the range.
+        A value of 0 or (0, 0) will result in no change.
+
+    Returns
+    -------
+    out : cupy.ndarray or numpy.ndarray
+        Output data. Same dimensions and type as input.
+
+    Raises
+    ------
+    ValueError
+        If 'brightness','contrast','saturation' or 'hue' is outside
+        of allowed range
+    TypeError
+        If input 'img' is not cupy.ndarray or numpy.ndarray
+
+    Examples
+    --------
+    >>> import cucim.core.operations.color as ccl
+    >>> # input is channel first 3d array
+    >>> output_array = ccl.color_jitter(input_arr,.25,.75,.25,.04)
+    """
+    try:
+        # should be a class stateful implementation to caches values
+        # once instead of checking every time
+        def check_input(value, name, center=1, bound=(0, float('inf')),
+                        clip_first_on_zero=True):
+            if isinstance(value, numbers.Number):
+                if value < 0:
+                    raise ValueError("If {} is a single number, \
+                                     it must be non negative.".format(name))
+                value = [center - float(value), center + float(value)]
+                if clip_first_on_zero:
+                    value[0] = max(value[0], 0.0)
+            elif isinstance(value, (tuple, list)) and len(value) == 2:
+                if not bound[0] <= value[0] <= value[1] <= bound[1]:
+                    raise ValueError("{} values should be between {}"
+                                     .format(name, bound))
+            else:
+                raise TypeError("{} should be a single number or a \
+                                list/tuple with length 2.".format(name))
+            # if value is 0 or (1., 1.) for brightness/contrast/saturation
+            # or (0., 0.) for hue, do nothing
+            if value[0] == value[1] == center:
+                value = None
+            return value
+
+        def get_params(brightness: Optional[List[float]],
+                       contrast: Optional[List[float]],
+                       saturation: Optional[List[float]],
+                       hue: Optional[List[float]]
+                       ) -> Tuple[np.ndarray, Optional[float],
+                                  Optional[float], Optional[float],
+                                  Optional[float]]:
```

Comment:
Is it necessary for these to be closures? Or could we define module level functions for these tasks?

Reply:
I propose to fix this in a separate MR.